Add deprecation messages for functorch.* function transforms

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92280
* __->__ #92279
* #92278

This PR:
- adds deprecation warnings when calling the functorch APIs
- adds documentation saying that those APIs are deprecated

It does this by creating thin wrappers around the original APIs that (1)
raise deprecation warnings and (2) have an additional line in their
documentation that they are deprecated.

NB:
- Python surpresses DeprecationWarning, so we use UserWarning instead.

Test Plan:
- New tests
- the functorch.* APIs are still tested for correctness because that's
what test/functorch/* use (as opposed to directly calling the
torch.func.* APIs)

zou3519 committed:
Add deprecation messages for functorch.* function transforms

This PR:
- adds deprecation warnings when calling the functorch APIs
- adds documentation saying that those APIs are deprecated

It does this by creating thin wrappers around the original APIs that (1)
raise deprecation warnings and (2) have an additional line in their
documentation that they are deprecated.

NB:
- PyTorch actually suppresses all DeprecationWarning by default.
So the warnings themselves are less important than the docs.

Test Plan:
- New tests
- the functorch.* APIs are still tested for correctness because that's
what test/functorch/* use (as opposed to directly calling the
torch.func.* APIs)

[ghstack-poisoned]

zou3519 committed:
Update on "Add deprecation messages for functorch.* function transforms"


This PR:
- adds deprecation warnings when calling the functorch APIs
- adds documentation saying that those APIs are deprecated

It does this by creating thin wrappers around the original APIs that (1)
raise deprecation warnings and (2) have an additional line in their
documentation that they are deprecated.

NB:
- Python surpresses DeprecationWarning, so we use UserWarning instead.

Test Plan:
- New tests
- the functorch.* APIs are still tested for correctness because that's
what test/functorch/* use (as opposed to directly calling the
torch.func.* APIs)

[ghstack-poisoned]

albanD commented on torch/_functorch/deprecated.py:
```diff
@@ -0,0 +1,107 @@
+
+def warn_deprecated(api, new_api=None):
+    warning = get_warning(api, new_api, replace_newlines=True)
+    warnings.warn(warning)
```
You might want to pass in stacklevel=2 so that the line number in the warning points to `functorch.vmap()` call and not `warn_deprecated('vmap', 'torch.vmap')`

zou3519 replied:
Good point

albanD commented on torch/_functorch/deprecated.py:
```diff
@@ -0,0 +1,107 @@
+    warnings.warn(warning)
+
+
+def setup_docs_annotations(functorch_api, torch_func_api=None, new_api_name=None):
```
Wouldn't adding `@functools.wraps()` to these function do most of what this function does?

zou3519 replied:
functools.wraps() propagates `__module__, __name__, __qualname__, __annotations__ and __doc__,`. Here we want to propagate `__annotations__` while adding a modified `__doc__` but none of the other attributes. The logic is different enough that I didn't use functools.wraps

albanD reviewed (approved):
Sounds good to me, just small details

soulitzer reviewed (approved):
LGTM

zou3519 committed:
Update on "Add deprecation messages for functorch.* function transforms"


This PR:
- adds deprecation warnings when calling the functorch APIs
- adds documentation saying that those APIs are deprecated

It does this by creating thin wrappers around the original APIs that (1)
raise deprecation warnings and (2) have an additional line in their
documentation that they are deprecated.

NB:
- Python surpresses DeprecationWarning, so we use UserWarning instead.

Test Plan:
- New tests
- the functorch.* APIs are still tested for correctness because that's
what test/functorch/* use (as opposed to directly calling the
torch.func.* APIs)

[ghstack-poisoned]

zou3519 committed:
Update on "Add deprecation messages for functorch.* function transforms"


This PR:
- adds deprecation warnings when calling the functorch APIs
- adds documentation saying that those APIs are deprecated

It does this by creating thin wrappers around the original APIs that (1)
raise deprecation warnings and (2) have an additional line in their
documentation that they are deprecated.

NB:
- Python surpresses DeprecationWarning, so we use UserWarning instead.

Test Plan:
- New tests
- the functorch.* APIs are still tested for correctness because that's
what test/functorch/* use (as opposed to directly calling the
torch.func.* APIs)

[ghstack-poisoned]

zou3519 committed:
Update on "Add deprecation messages for functorch.* function transforms"


This PR:
- adds deprecation warnings when calling the functorch APIs
- adds documentation saying that those APIs are deprecated

It does this by creating thin wrappers around the original APIs that (1)
raise deprecation warnings and (2) have an additional line in their
documentation that they are deprecated.

NB:
- Python surpresses DeprecationWarning, so we use UserWarning instead.

Test Plan:
- New tests
- the functorch.* APIs are still tested for correctness because that's
what test/functorch/* use (as opposed to directly calling the
torch.func.* APIs)

[ghstack-poisoned]
